### ISSUE TITLE
SFTP still not fully working: large files upload

### ISSUE THREAD

**Issue by MarvinFS** [owncloud/core#9632]
Regarding latest  #9605 SFTP fixes, i'm trying the latest nightly 
https://download.owncloud.org/community/daily/owncloud-daily-master.tar.bz2
new DB new everything

### Steps to reproduce
1. upload multiple files
2. download multiple files
3.  upload large file (1.5GB, 4GB)

### Expected behaviour
1. files are uploaded 
2. successful download
3. successful large file upload

### Actual behaviour
1. working just fine - fixed now, speed is also great with SFTP
2. also everything ok here!
3. while the progress bar fills in pretty fast (depends on network speed)
the actual files are failed to upload either 1.5GB or 4GB with several errors in logs.
files ends up partially uploaded...

### Server configuration
Ubuntu 14LTS

**Web server:**
Apache 2.4

**Database:**
MySQL

**PHP version:**
PHP 5.5.9-1ubuntu4.3 

**ownCloud version:** (see ownCloud admin page)
july 15 nightly

**Updated from an older ownCloud or fresh install:**
the very fresh

**List of activated apps:**
only defaults and external storages app

**The content of config/config.php:**
```
<?php
$CONFIG = array (
  'instanceid' => 'occ5e8b3c49a',
  'passwordsalt' => '',
  'trusted_domains' =>
  array (
    0 => '10.0.0.37',
  ),
  'datadirectory' => '/var/www/owncloud/data',
  'dbtype' => 'mysql',
  'version' => '7.8.0.0',
  'dbname' => '',
  'dbhost' => 'localhost',
  'dbtableprefix' => '',
  'dbuser' => '
  'dbpassword' => '',
  'installed' => true,
  'ldapIgnoreNamingRules' => false,
  'loglevel' => '0',
  'forcessl' => true,
  'logtimezone' => 'Asia/Yekaterinburg',
  'log_authfailip' => true,
  'log_rotate_size' => 10485760,
  'debug' => true,
  'xframe_restriction' => true,
  'theme' => '',
  'maintenance' => false,
);
```

**Are you using external storage, if yes which one:** 
sftp

**Are you using encryption:** yes/no
no

### Client configuration
**Browser:**
firefox, ie11, chrome
**Operating system:**
windows 7, 8.1

#### ownCloud log (data/owncloud.log)
```
{"reqId":"53c4b8bacf667","app":"OC\\Files\\Cache\\Scanner","message":"!!! No reuse of etag for 'files_versions\/WebDAV\/test4' !!! \ncache: Array\n(\n    [fileid] => 89\n    [storage] => home::admin\n    [path] => files_versions\/WebDAV\/test4\n    [parent] => 88\n    [name] => test4\n    [mimetype] => httpd\/unix-directory\n    [mimepart] => httpd\n    [size] => 0\n    [mtime] => 1405401121\n    [storage_mtime] => 1405401121\n    [encrypted] => \n    [unencrypted_size] => 0\n    [etag] => 53c4b823195f0\n    [permissions] => 31\n)\n \ndata: Array\n(\n    [mimetype] => httpd\/unix-directory\n    [mtime] => 1405401125\n    [size] => -1\n    [etag] => 53c4b8bacf331\n    [storage_mtime] => 1405401125\n    [permissions] => 31\n    [parent] => 88\n)\n","level":0,"time":"2014-07-15T11:14:34+06:00","method":"GET","url":"\/owncloud\/index.php\/apps\/files\/ajax\/scan.php?force=false&dir=&requesttoken=f06feb01eac59e487e78"}
{"reqId":"53c4bdd02c1f8","app":"OC\\Files\\Cache\\Scanner","message":"!!! No reuse of etag for 'ocloud\/test4' !!! \ncache: Array\n(\n    [fileid] => 27\n    [storage] => sftp::ocloud@marvinfs.convex.ru\/\/\n    [path] => ocloud\/test4\n    [parent] => 17\n    [name] => test4\n    [mimetype] => httpd\/unix-directory\n    [mimepart] => httpd\n    [size] => 80245\n    [mtime] => 1405400671\n    [storage_mtime] => 1405400671\n    [encrypted] => \n    [unencrypted_size] => 0\n    [etag] => 53c4b6104fe13\n    [permissions] => 31\n)\n \ndata: Array\n(\n    [mimetype] => httpd\/unix-directory\n    [mtime] => 1405402518\n    [size] => -1\n    [etag] => 53c4bdd026181\n    [storage_mtime] => 1405402518\n    [permissions] => 31\n    [parent] => 17\n)\n","level":0,"time":"2014-07-15T11:36:16+06:00","method":"GET","url":"\/owncloud\/index.php\/apps\/files\/ajax\/getstoragestats.php?dir=%2FSFTP%2Focloud%2Ftest4"}
{"reqId":"53c4c0e5778f9","app":"webdav client","message":"Not authenticated","level":3,"time":"2014-07-15T11:49:25+06:00","method":"GET","url":"\/owncloud\/index.php\/apps\/files\/ajax\/scan.php?force=false&dir=&requesttoken=f06feb01eac59e487e78"}
{"reqId":"53c4c3b70df28","app":"PHP","message":"Connection closed prematurely at \/var\/www\/owncloud\/3rdparty\/phpseclib\/phpseclib\/phpseclib\/Net\/SSH2.php#2579","level":3,"time":"2014-07-15T12:01:27+06:00","method":"POST","url":"\/owncloud\/index.php\/apps\/files\/ajax\/upload.php"}
{"reqId":"53c4c3b70df28","app":"PHP","message":"Connection closed by server at \/var\/www\/owncloud\/3rdparty\/phpseclib\/phpseclib\/phpseclib\/Net\/SSH2.php#2825","level":3,"time":"2014-07-15T12:01:27+06:00","method":"POST","url":"\/owncloud\/index.php\/apps\/files\/ajax\/upload.php"}
{"reqId":"53c4c3b70df28","app":"PHP","message":"Expected SSH_FXP_STATUS at \/var\/www\/owncloud\/3rdparty\/phpseclib\/phpseclib\/phpseclib\/Net\/SFTP.php#1590","level":3,"time":"2014-07-15T12:01:27+06:00","method":"POST","url":"\/owncloud\/index.php\/apps\/files\/ajax\/upload.php"}
{"reqId":"53c4c3b70df28","app":"PHP","message":"Connection closed prematurely at \/var\/www\/owncloud\/3rdparty\/phpseclib\/phpseclib\/phpseclib\/Net\/SSH2.php#2992","level":3,"time":"2014-07-15T12:01:27+06:00","method":"POST","url":"\/owncloud\/index.php\/apps\/files\/ajax\/upload.php"}
{"reqId":"53c4c3b70df28","app":"OC\\Files\\Cache\\Scanner","message":"!!! No reuse of etag for 'ocloud\/test4\/Delo.Rumjanceva.1955.DVDRip_[Youtracker]_by_AVP Studio.avi' !!! \ncache: Array\n(\n    [fileid] => 174\n    [storage] => sftp::ocloud@marvinfs.convex.ru\/\/\n    [path] => ocloud\/test4\/Delo.Rumjanceva.1955.DVDRip_[Youtracker]_by_AVP Studio.avi\n    [parent] => 27\n    [name] => Delo.Rumjanceva.1955.DVDRip_[Youtracker]_by_AVP Studio.avi\n    [mimetype] => \n    [mimepart] => video\n    [size] => 66953216\n    [mtime] => 1405402656\n    [storage_mtime] => 1405402656\n    [encrypted] => \n    [unencrypted_size] => 0\n    [etag] => 53c4bdd0501bf\n    [permissions] => 27\n)\n \ndata: Array\n(\n    [mimetype] => video\/x-msvideo\n    [mtime] => 1405404166\n    [size] => 817790918\n    [etag] => 53c4c3b81f14a\n    [storage_mtime] => 1405404166\n    [permissions] => 27\n    [parent] => 27\n)\n","level":0,"time":"2014-07-15T12:01:28+06:00","method":"POST","url":"\/owncloud\/index.php\/apps\/files\/ajax\/upload.php"}
{"reqId":"53c4c3b70df28","app":"core","message":"Exception during mimetype insertion: An exception occurred while executing 'INSERT INTO `oc_mimetypes`(`mimetype`) VALUES(?)':\n\nSQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'video\/x-msvideo' for key 'mimetype_id_index'","level":0,"time":"2014-07-15T12:01:28+06:00","method":"POST","url":"\/owncloud\/index.php\/apps\/files\/ajax\/upload.php"}

```
php options in .htaccess

```
php_value upload_max_filesize 16G
php_value post_max_size 16G
php_value memory_limit 1024M
php_value output_buffering 16384
php_value mbstring.func_overload 0
php_value max_input_time 36484
php_value max_execution_time 36484

```

**Comment by DeepDiver1975** [comment]
@bantu is there anything more we can do about this issue? Did we already update phpseclib properly? THX

**Comment by nickvergessen** [comment]
Well we are on 0.3.6 (Feb 2014) and current is 0.3.9 (Nov 2014)

**Comment by DeepDiver1975** [comment]
on stable7 we are hopelessly out dated with respect to your 3rdparty dependencies: https://www.versioneye.com/user/projects/54bf83f36c0035c592000048

master/oc8 doesn't look that bad: https://www.versioneye.com/user/projects/54a548e697427553100000b9

So from my understanding we should update phpseclib for stable7
@bantu can I ask you to update this in the 3rdparty repo stable7 and apply proper testing? Many thanks

**Comment by DeepDiver1975** [comment]
@bantu ping

**Comment by bantu** [comment]
@DeepDiver1975 Not sure there is anything that can be done about this ticket in particular, but if PHP 5.6.1+ compatibility is desired (I think it is), we should upgrade to 0.3.9 to get https://github.com/phpseclib/phpseclib/pull/493/files
Besides that, the improvements to Net_SSH2 are important for our SFTP support.
Also see https://github.com/phpseclib/phpseclib/blob/master/CHANGELOG.md

**Comment by bantu** [comment]
@MarvinFS Could you please tell us a bit more about your SFTP server? Which operating system is it running. Is it running OpenSSH? Which version?

**Comment by nickvergessen** [comment]
@bantu you dont think updating seclib from 0.3.6 to 0.3.9 will help?

**Comment by bantu** [comment]
@nickvergessen I don't know. stable7 should receive a phpseclib upgrade for PHP 5.6 compatibility. @MarvinFS here claims to be using master, which already has a newer version of phpseclib (at least right now).

**Comment by DeepDiver1975** [comment]
just to get a clear understanding of the current situation:
on master we can upload files to sftp with a max size of ???
on stable7 we can upload files to sftp with a max size of ???

**Comment by nickvergessen** [comment]
@bantu master uses 0.3.6 I just verified it in the repo

**Comment by bantu** [comment]
@nickvergessen Then https://www.versioneye.com/user/projects/54a548e697427553100000b9 lies.

**Comment by nickvergessen** [comment]
@bantu appearantly it does, maybe it's confused by the "*" requirement we have.
Current 3rdparty master: https://github.com/owncloud/3rdparty/blob/master/composer.lock#L1278-L1279
Current commit of oc master: https://github.com/owncloud/3rdparty/blob/a32d3924bd0012a5410fff4666131cbdfdec2001/composer.lock#L1278-L1279

**Comment by nickvergessen** [comment]
Also:
```
nickv@nickv-lt-mint ~/ownCloud/Master/3rdparty $ ~/Tools/composer.phar update phpseclib/phpseclib
Loading composer repositories with package information
Updating dependencies (including require-dev)
  - Removing phpseclib/phpseclib (0.3.6)
  - Installing phpseclib/phpseclib (0.3.9)
    Downloading: 100%         

Writing lock file
Generating autoload files

**Comment by bantu** [comment]
@DeepDiver1975 I've not actually interpreted this ticket to be about addressing any potentially remaining filesize limitations. Rather I've always focussed on the "Expected SSH_FXP_STATUS" / "Connection closed prematurely" errors, which were encountered when uploading large files (say 1.8 GiB). I've not been able to reproduce those errors so far.

The version of phpseclib stable7 is using does not clamp filesizes to 4 GiB any more, as per https://github.com/owncloud/core/issues/11432#issuecomment-59217651. As such, there should be no practical filesize limit on 64bit environments.

On 32bit system the situation is a bit different. A call to $sftp->put() will result in filesize() being called by phpseclib, which as we know might lie. A solution to this part is to extract the filesize tricks into its own library and also use that in phpseclib. I've been meaning to do that.

As per nickvergessen, stable7 and master currently use the same version.

**Comment by bantu** [comment]
@DeepDiver1975 Additionally, I believe the web interface can not upload files larger than 2 GiB at all on PHP versions below 5.6 (see http://cweiske.de/tagebuch/php-large-file-uploads.htm). And chunking still needs to be investigated on 32bit platforms.

**Comment by MarvinFS** [comment]
Guys, i'm still greatly concerned about this issue, thank you very much for not letting it go, but i'm on a ski resort right now, will be back on sunday evening and from monday's morning i'll be able to test any new builds (if any) and will provide all the info requested, ok?

**Comment by MorrisJobke** [comment]
@DeepDiver1975 From IRC:

> bantu: move to next version I'd say. I haven't been able to reproduce this problem in particular.

Move to 8.1 then

**Comment by MarvinFS** [comment]
Colleagues, what particular test build i may use to see if these issues are still exists? 
I'm ready to test 'em with latest debian and ubuntu LTS
including sftp transfer and large files upload

**Comment by MarvinFS** [comment]
@bantu i've just tried again following your instructions (replace the contents of the owncloud/3rdparty/phpseclib/phpseclib folder of https://download.owncloud.org/community/daily/owncloud-daily-master.tar.bz2 with the contents of the phpseclib-master folder of https://github.com/phpseclib/phpseclib/archive/master.zip which is 0.3.9 apparently ) and in doesn't work anymore - 
ocloud is not starting - giving me 

{"reqId":"dmJdtH1soUKmfzuP3l3E","remoteAddr":"192.168.1.101","app":"PHP","message":"Class 'phpseclib\\Crypt\\Rijndael' not found at \/var\/www\/owncloud\/3rdparty\/phpseclib\/phpseclib\/phpseclib\/Crypt\/AES.php#58","level":3,"time":"2015-02-10T22:43:59+05:00"}

with original phpseclib from latest ocloud daily master the behavior is still the same - uploads about 818MB to the SFTP server and disconnects leaving partially uploaded file

**Comment by bantu** [comment]
@MarvinFS Branches in phpseclib have been reorganised. You will have to use https://github.com/phpseclib/phpseclib/archive/1.0.zip instead.

**Comment by MarvinFS** [comment]
@bantu 
ok i finished the testing with new 0.3.9 branch (all the testing were over public internet)
source server: at office, vmware cluster, ubuntu 14.04 LTS: ocloud daily master clean + 0.3.9. phpaseclib (external storage support app enabled)

dest server: at home, debian 7.8

upload 2.5GB file from source to dest by SFTP external storage - SUCCESS speed 1.3-1.6MB/sec
upload 4.5GB file from source to dest by FTP external storage - SUCCESS speed 11.2 - 11.5 MB\sec

1. the low speed for SFTP upload is due to speed limitations for openssh and i failed to find any suitable repo for HPN-SSH for Debian (found only for Ubuntu)
2. the FTP upload is just fine, although it was good working in the past...  
3. 1st issue: when you start to upload the file to external storage in seems that first it uploads to the ocloud server ifself (the progress bar fills from 0 to full), during this time the dest server is IDLE!!! only after the progress bar fills to full actual upload begins to the dest. server. During the actual upload the progress bar stays full until the upload is finished. But as it's an external storage the actual file is not held on ocloud server itself - so it's not clear what it's doing during initial progress bar fill... 
4. 2nd issue: for the FTP upload the file on dest server is complete and full, but inside ocloud interface, the file size shows as 394MB only... not sure why... refreshing the folder don't help... (screenshot https://yadi.sk/i/w__K8zYQebN98 )

in the logfiles nothing useful  - seems everything fine.

Whay else can i do?

**Comment by MarvinFS** [comment]
you will be laughing but now if i try to download that uploaded file from FTP server though ocloud interface it downloads ONLY the ocloud size which is 359.4 MB, if i try to download the same file from remote FTP server directly is downloads the whole 4.5 gigs... md5 matches - i checked it

**Comment by PVince81** [comment]
@MarvinFS if you uploaded through the Web UI that goes through POST, in which case PHP will first store a temp file before giving OC access to it. This might explain the delay.
If you try with WebDAV it should be streamed directly.

**Comment by MarvinFS** [comment]
@PVince81 Got it! makes sense! but the ocloud interface FTP file size issue after upload still remains - should i open other tread? it's not related to phpseclib due to this is plain FTP.

now will try the maximum size upload through SFTP using HPN-SSH on ubuntu - such low transfer speed of standard openssh drives me nuts

**Comment by PVince81** [comment]
Yes, please make another ticket for FTP. Thanks.

**Comment by MarvinFS** [comment]
uploaded 9.2GB file to SFTP external storage just fine! success at last! 
next will try 16G.
i used HPN-SSH as a SFTP external target server - performance improved DRASTICALLY - 5-10 times! (from 1.5 MB\sec to 5-1MB\sec) i used this manual for Ubuntu https://spoutcraft.org/threads/blazing-fast-sftp-ssh-transfer.7682/
although i failed to find a ready repo for Debian

**Comment by MarvinFS** [comment]
succeeded with 14GB file SFTP upload! it took about 1.5 hours (5-7MB/sec with HPN-SSH)
working fine, i guess 0.3.9 solved the issue
are we closing the ticket?

**Comment by bantu** [comment]
@MarvinFS Yeah, let's close it. Closing this in favour of #12115

**Comment by bantu** [comment]
@MarvinFS Thanks a lot for your patience and testing.
